File: Leetcode/TargetSum.cpp
class Solution {
   public:
    int findTargetSumWays(vector<int>& nums, int S) {
        if (nums.empty()) return S == 0;
        int mx = 2000;
        if (S > mx) return 0;

        vector<vector<int>> dp(nums.size(), vector<int>(4002));
        dp[0][mx + nums[0]]++;
        dp[0][mx - nums[0]]++;

        for (int i = 1; i < nums.size(); ++i) {
            for (int k = nums[i]; k <= 2 * mx - nums[i]; ++k) {
                dp[i][k] += dp[i - 1][k - nums[i]];
                dp[i][k] += dp[i - 1][k + nums[i]];
            }
        }

        return dp[nums.size() - 1][mx + S];
    }
};

File: Leetcode/TargetSum.rs
// https://leetcode.com/problems/target-sum

use std::collections::HashMap;

struct Solution {}

impl Solution {
    pub fn find_target_sum_ways(nums: Vec<i32>, target: i32) -> i32 {
        let mut dp = HashMap::<i32, i32>::new();
        dp.insert(0, 1);

        for num in nums {
            let mut ndp = HashMap::<i32, i32>::new();
            for (&have_num, &cnt) in dp.iter() {
                *ndp.entry(have_num - num).or_default() += cnt;
                *ndp.entry(have_num + num).or_default() += cnt;
            }

            dp = ndp;
        }

        *dp.get(&target).unwrap_or(&0)
    }
}
